### ISSUE TITLE
Text sometimes entered right-to-left in chunks

### ISSUE THREAD

**Issue by rmcd1024** [polymode/polymode#226]
Sometimes when editing within chunks with polymode activated, Emacs reverses text, so what I type is entered right to left instead of left-to-right. This behavior stops when I leave the chunk. I have posted a detailed question and steps to reproduce on [emacs.stackexchange](https://emacs.stackexchange.com/questions/50323/text-entered-right-to-left-in-chunks-using-polymode). 

By the way, thanks for polymode!

**Comment by vspinu** [comment]
I don't see this, sorry. This would have been an obvious bug. So it's must be something local.

It's most likely an interaction with some other minor mode which you have. You could disable them in turn and see which one is causing the problem. 

Ideally you should be able to reproduce this from poly-markdown repo with `make start`. This will start a clean emacs only with poly-markdown + dependencies loaded.

**Comment by rmcd1024** [comment]
Reproducing requires that `poly-r` and `ess` also be loaded. I can reproduce as follows:

1. `make start` in the poly-markdown git repo
2. within emacs (26.1 in my case) `package-list-packages` and install `ess` and `poly-r`, both latest versions (both 2019-05) (installing `ess` is necessary for the next step to have the intended effect)
3. execute this lisp: `(add-to-list 'auto-mode-alist '("\\.Rmd" . poly-markdown+r-mode))` 
4. Follow the keystrokes in https://emacs.stackexchange.com/questions/50323/text-entered-right-to-left-in-chunks-using-polymode. 

When you visit `test.Rmd` in step 4 the status line should show Markdown and PM-Rmd as active modes. 

For me, steps 1-4 reproduce the bug. If I don't install ess and poly-r, there's no problem.

**Comment by vspinu** [comment]
Thanks. Which repo do you `make start`, polymode? Steps 2 and 3 are not necessary if you do `make start` from `poly-r` repo. Could you please try? I cannot reproduce with emacs 26.1. 

Just open new file test.Rmd and type 

```
---         ;; self-insert-command * 3
RET         ;; markdown-enter-key
---         ;; self-insert-command * 3
C-a         ;; move-beginning-of-line
RET         ;; markdown-enter-key
C-p         ;; previous-line
123         ;; self-insert-command * 3
```
right?

**Comment by vspinu** [comment]
You know, the strange part is that with `make start` yaml is not there so you basically see the inversion in plain markdown, which is more than weird :/.

**Comment by vspinu** [comment]
Nevermind. I see it in `poly-markdown` repo. You probably installed yaml by hand.

**Comment by vspinu** [comment]
This doesn't come from polymode. Looks like something very low level. I am disabling polymode stuff entirely , post-comand hooks, font-lock and after-change and still seeing it. 

The good news is that it really occurs only in your specific test case. If you revert the buffer it's no longer there.

**Comment by rmcd1024** [comment]
It's great that you could reproduce, thanks. Regarding the likelihood of this occurring, I actually run into this quite frequently. It can occur with R chunks as well. It seems to occur if I've moved or manipulated the closing chunk boundary. I'm surprised that this hasn't popped up before. I didn't try to create a second reproducible case but I can try if it would be useful.

Otherwise, I'm guessing that more from me wouldn't be helpful, is that correct?

**Comment by vspinu** [comment]
I actually tracked it down to `post-self-insert-hook`. There is a compiled function there which I don't know the origin and cannot find which mode places it in there. If you can reproduce with other innermodes can you posthere what you see in `post-self-insert-hook`?

**Comment by vspinu** [comment]
Ok, I figured out the culprit. It's the ugly code in `newline` which does some crazy hackery with post-self-insert-hook. I might need to bring this to emacs devel.

**Comment by rmcd1024** [comment]
Very nice! If you do need me to do more, I am willing but will need a pointer to help me learn how to debug emacs lisp.

**Comment by rmcd1024** [comment]
Here is a key sequence that works with an R chunk. It's slightly different than with the yaml chunk. I enter 123 twice and it only is reversed the second time. The key thing here (I think) is to go to the end of chunk, beginning of line, backspace when there's content in the chunk, then hit return, then previous-line, then it enters the weird reverse mode.

```			;; self-insert-command * 3
{			;; skeleton-pair-insert-maybe
r			;; self-insert-command
}			;; skeleton-pair-insert-maybe
2*RET			;; newline
```			;; self-insert-command * 3
C-p			;; previous-line
123			;; self-insert-command * 3
C-n			;; next-line
C-a			;; move-beginning-of-line
DEL			;; delete-backward-char
2*RET			;; newline
C-p			;; previous-line
123			;; self-insert-command * 3

**Comment by vspinu** [comment]
Should be fixed. Could you please try? I am advising `newline` but this should be really fixed on emacs side. Thanks for reporting. I am really amazed how this went unnoticed so far :/

**Comment by rmcd1024** [comment]
Thanks so much! It appears to be fixed with polymode-20190507.
